src/VisualStudio/Core/Def/Implementation/Utilities/ServiceInitializer.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Roslyn.Utilities;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Utilities
+{
+    internal class ServiceInitializer<TInterface, TService> where TInterface : class
+    {
+        private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
+
```

Comment:
The lock above, and the code that tries to initialize things here is fairly scary.  @jasonmalinowski could this just be replaced with AsyncLazy? It feels like tht provides the abstraction we need. 

1. it can be both async or sync.
2. it only lets one person be computing. 
3. it can be setup to cache the result on success.

Is there a reason AsyncLazy would not be the preferred route here?

Reply:
AsyncLazy doesn't use JTF to wait.

Reply:
Why do you need JTF?

Reply:
GetServiceAsync assumes JTF to work properly.

Reply:
The JTF used by `GetServiceAsync` is *not* a substitute for your own use of JTF where you block the main thread. This is still vulnerable to race conditions, which once you fix, it will then be vulnerable to deadlocks unless you do the JTF dance properly. The *vs-threading* version of `AsyncLazy<T>` *will* fix this all for you (and your sync blocking methods will need to use JTF.Run while calling `AsyncLazy<T>.GetValueAsync()`. I highly recommend you switch to it.